verilog/analysis/checkers/forbid_implicit_declarations_rule.h

```diff
@@ -0,0 +1,133 @@
+// Copyright 2017-2020 The Verible Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef VERIBLE_VERILOG_ANALYSIS_CHECKERS_FORBID_IMPLICIT_DECLARATIONS_RULE_H_
+#define VERIBLE_VERILOG_ANALYSIS_CHECKERS_FORBID_IMPLICIT_DECLARATIONS_RULE_H_
+
+#include <set>
+#include <string>
+
+#include "common/analysis/lint_rule_status.h"
+#include "common/analysis/matcher/core_matchers.h"
+#include "common/analysis/matcher/matcher.h"
+#include "common/analysis/matcher/matcher_builders.h"
+#include "common/analysis/syntax_tree_lint_rule.h"
+#include "common/text/symbol.h"
+#include "common/text/syntax_tree_context.h"
+#include "verilog/CST/verilog_matchers.h"  // IWYU pragma: keep
+#include "verilog/analysis/descriptions.h"
+
+namespace verilog {
+namespace analysis {
+
+// ForbidImplicitDeclarationsRule detect implicitly declared nets
+class ForbidImplicitDeclarationsRule : public verible::SyntaxTreeLintRule {
```

Comment:
Ok, this is my first serious look at the implementation of this analysis, after we've amassed a healthy collection of test cases.
I'll start with high-level design thoughts.  @hzeller This might be of interest to you, since you're looking at lint rule design issues these days.

I see that you've chosen to base this on `SyntaxTreeLintRule` which _seems_ intuitive, and a combination of matchers.  

* Keep in mind that matchers, though powerful and expressive, run on every node that is visited (`HandleNode`), so every node is starting a new pattern search.
* The limitation of a using `SyntaxTreeLintRule` is that it lacks _control_ over the visitation of nodes, which is being done in `SyntaxTreeLinter::Visit`.  Why was this done?  `SyntaxTreeLinter` collects a bunch of rules, and executes them all in a _single CST traversal_ that builds up and tears down SyntaxTreeContext _once_.  The implication is that each instantiated rule object must maintain enough state to have its actions _interleaved_ among other "concurrently" executing rules.  Had we let each rule manage its own traversal, they could simply look like functions of CST structures that return collections of lint violations.  However, we are not forced to share traversals with a `SyntaxTreeLintRule`, we have the option of owning and customizing traversal by using the more general `TextStructureLintRule`.

Looking over the current implementation, I see effort in maintaining a stack that resembles the `SyntaxTreeContext` maintained by `SyntaxTreeLinter` and its base `TreeContextVisitor`.  (Note that `verible::TreeUnwrapper` is another subclass of `TreeContextVisitor`.)  Only instead of growing the stack on every node, you only grow the stack on _nodes bearing lexical scope_.  If you extended `TreeContextVisitor` to include a scope-stack (Verilog-specific, no longer language-agnostic) in addition to the normal `SyntaxTreeContext` stack, that could be the basis for **all scoped analyses** that use symbol tables (like your `DeclType`).  

What about matching?  In past PRs I've often offered the alternative of using direct substructure access through CST/ functions, which is much faster, because it is not based on any tree-searching/predicate-matching.  The alternative to grabbing `.Bind()` symbols from `BoundSymbolManager` is direct access.

* From every net declaration (or element declaration) you can directly reach the identifier(s) it declares precisely.
* From every assignment you can isolate the LHS from the RHS:
  * You could set state variables to distinguish when you're in the left vs. right instead of trying to infer from upward inspection.  
  * Based on left vs. right, you would know whether you are looking at lvalues vs. rvalues (but LHS references can have rvalues in array indices, so be cautious there.)
* Most of the logic would be in the form of a sparse switch-case statement on node types (much like the formatter's `TreeUnwrapper::SetIndentationsAndCreatePartitions`) whose interesting cases include declarations of several types, assignments, instantiations, loop/conditional generate blocks, references, and non-interesting cases would just be default traversal.
 
Does this sound simpler than your current implementation?